rclpy/src/rclpy/_rclpy.c

```diff
@@ -3369,6 +3370,385 @@ rclpy_clock_set_ros_time_override(PyObject * Py_UNUSED(self), PyObject * args)
+        return false;
+      }
+      /* This was a borrowed reference. INCREF'd so we can unconditionally DECREF below. */
+      Py_INCREF(parameter_dict);
+    }
+    rcl_node_params_t node_params = params->params[i];
+    for (size_t ii = 0; ii < node_params.num_params; ii++) {
+      PyObject * py_param_name = PyUnicode_FromString(node_params.parameter_names[ii]);
+      if (NULL == py_param_name) {
+        Py_DECREF(py_node_name);
+        Py_DECREF(parameter_dict);
+        return false;
+      }
+      PyObject * py_param = _parameter_from_rcl_variant(py_param_name,
+          &node_params.parameter_values[ii], parameter_cls, parameter_type_cls);
+      if (NULL == py_param) {
+        Py_DECREF(py_node_name);
+        Py_DECREF(parameter_dict);
+        Py_DECREF(py_param_name);
+        return false;
+      }
+      if (-1 == PyDict_SetItem(parameter_dict, py_param_name, py_param)) {
+        Py_DECREF(py_node_name);
+        Py_DECREF(py_param_name);
+        Py_DECREF(parameter_dict);
+        Py_DECREF(py_param);
+        return false;
+      }
+      Py_DECREF(py_param_name);
+      Py_DECREF(py_param);
+    }
+    Py_DECREF(py_node_name);
+    Py_DECREF(parameter_dict);
+  }
+  return true;
+}
+
+/// Populate a Python dict with node parameters parsed from arguments files
+/**
+ * On failure a Python exception is raised and false is returned if:
+ *
+ * Raises RuntimeError if param_files cannot be extracted from arguments.
+ * Raises RuntimeError if yaml files do not parse succesfully.
+ *
+ * \param[in] args The arguments to parse for parameter files
+ * \param[in] allocator Allocator to use for allocating and deallocating within the function.
+ * \param[out] params_by_node_name A Python dict object to place parsed parameters into.
+ *
+ * Returns true when parameters are parsed successfully (including the trivial case)
+ *         false when there was an error during parsing and a Python exception was raised.
+ *
+ */
+static bool
+_parse_param_files(
+  const rcl_arguments_t * args, rcl_allocator_t allocator, PyObject * parameter_cls,
+  PyObject * parameter_type_cls, PyObject * params_by_node_name)
+{
+  char ** param_files;
+  int param_files_count = rcl_arguments_get_param_files_count(args);
+  rcl_ret_t ret;
+  bool successful = true;
+  if (param_files_count > 0) {
+    ret = rcl_arguments_get_param_files(args, allocator, &param_files);
+    if (RCL_RET_OK != ret) {
+      PyErr_Format(PyExc_RuntimeError, "Failed to get initial parameters: %s",
+        rcl_get_error_string_safe());
+      return false;
+    }
+    for (int i = 0; i < param_files_count; i++) {
+      if (successful) {
+        rcl_params_t * params = rcl_yaml_node_struct_init(allocator);
+        if (!rcl_parse_yaml_file(param_files[i], params)) {
+          /* failure to parse will automatically fini the params struct */
+          PyErr_Format(PyExc_RuntimeError, "Failed to parse yaml params file: %s", param_files[i]);
+          successful = false;
```

Comment:
should we just cleanup and break out of the loop in this case?

Reply:
I think this might be a matter of preference, and I'm flexible. The current loop deallocates as it goes. We could either preserve that and have a short circuit finish cleanup with an inner for loop before returning or instead of cleaning as we go iterate a second time to do cleanup and break to that case on failure. I tend to prefer iterating a minimal number of times but the *n* of parameter files is potentially never going to be large enough that we notice *2n* > *n*.

Reply:
Yeah as the loop doesn't do anything else than cleaning up as soon a a failure occurs it's fine to keep as is :+1:

---

> We could either preserve that and have a short circuit finish cleanup with an inner for loop before returning or instead of cleaning as we go iterate a second time to do cleanup and break to that case on failure

Just to clarify, I was suggesting to clean up the "remaining ones" and breaking, not to change "the cleanup as we go" and requiring to iterate again over the entire set